File: src/bpm/bpm.controller.ts
import { Controller, Get, Query } from '@nestjs/common';
import { BpmService } from './bpm.service';

@Controller('bpm-lookup')
export class BpmController {
  constructor(private readonly bpmService: BpmService) {}

  @Get()
  async lookup(
    @Query('title') title: string,
    @Query('artist') artist: string,
  ): Promise<{ bpm: number | null }> {
    const bpm = await this.bpmService.lookup(title ?? '', artist ?? '');
    return { bpm };
  }
}

File: spring-src/main/java/com/blackout/api/bpm/infrastructure/web/BpmController.java
package com.blackout.api.bpm.infrastructure.web;

import com.blackout.api.bpm.application.service.BpmApplicationService;
import com.blackout.api.bpm.infrastructure.web.dto.BpmLookupRequest;
import com.blackout.api.bpm.infrastructure.web.dto.BpmLookupResponse;
import jakarta.validation.Valid;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/bpm")
public class BpmController {

    private final BpmApplicationService service;

    public BpmController(BpmApplicationService service) {
        this.service = service;
    }

    @PostMapping("/lookup")
    public BpmLookupResponse lookup(@Valid @RequestBody BpmLookupRequest request) {
        Integer bpm = service.lookup(request.title(), request.artist());
        return new BpmLookupResponse(bpm);
    }
}
